src/cluster.c

```diff
@@ -6837,10 +6837,10 @@ void clusterRedirectClient(client *c, clusterNode *n, int hashslot, int error_co
  * returns 1. Otherwise 0 is returned and no operation is performed. */
 int clusterRedirectBlockedClientIfNeeded(client *c) {
     if (c->flags & CLIENT_BLOCKED &&
-        (c->btype == BLOCKED_LIST ||
-         c->btype == BLOCKED_ZSET ||
-         c->btype == BLOCKED_STREAM ||
-         c->btype == BLOCKED_MODULE))
+        (c->bstate.btype == BLOCKED_LIST ||
+         c->bstate.btype == BLOCKED_ZSET ||
```

Comment:
As far as I can tell BLOCKED_LIST and BLOCKED_ZSET aren't handled separately, and they could be generalized to `BLOCKED_NO_KEY`. There is the ongoing BEVAL conversation, https://github.com/redis/redis/issues/9925, which we could trivially add support for with this.

Reply:
This is somewhat problematic. Let me explain why.
Currently we do not process command unless it was realy unblocked by the currently executed command.
when we try to unblock clients on the ready key we first match the blocked clients blocking type to the type of the key
``
if ((o != NULL && (receiver->bstate.btype == getBlockedTypeByType(o->type)))
``
so in case a client is blocked on mylist key for example it should not be reprocessed in case some other client will do ``ZADD mylist 1 1``
 for example...
 I guess we can think of improvement here which will also require us to maintain the blocking timeout differently.

Reply:
I will say I never really liked this optimization to begin with, since it's very broad and assumes there is only one type of operations being done on the dataset. I'm ok forgoing it, but maybe we can make a note to find more optimizations. I don't think this part of the codebase has been optimized much for performance.

Reply:
I agree that a better infrastructure would not need to be aware of a specific blocking-on-keys type, but rather  a general "blocked-on-keys", I also agree we would like to have more blocking types in the future.
However this will require much deeper changes since the blocking timeout should also be refactored, In case we will allow reprocessing a command in order to evaluate if it is still blocked. Not a huge change but it will probably complicate some cases.